symengine/expand.cpp

```diff
@@ -0,0 +1,317 @@
+                    term = Mul::from_dict(one, std::move(d2));
+                }
+                Add::dict_add_term(d_, _mulnum(multiply, coef2), term);
+            }
+        }
+    }
+
+    void bvisit(const Pow &self) {
+        RCP<const Basic> _base = expand(self.get_base());
+
+        if (is_a<Integer>(*self.get_exp()) && is_a<UnivariatePolynomial>(*_base)) {
+            int q = rcp_static_cast<const Integer>(self.get_exp())->as_int();
+            RCP<const UnivariatePolynomial> p = rcp_static_cast<const UnivariatePolynomial>(_base);
+            RCP<const UnivariatePolynomial> r = univariate_polynomial(p->var_, 0, {{0, 1}});
+            while (q != 0) {
+                if (q % 2 == 1) {
+                    r = mul_uni_poly(r, p);
+                    q--;
+                }
+                p = mul_uni_poly(p, p);
+                q /= 2;
+            }
+            _coef_dict_add_term(multiply, r);
+            return;
+        }
+
+        if (!is_a<Integer>(*self.get_exp()) || !is_a<Add>(*_base)) {
+            if (neq(*_base, *self.get_base())) {
+                Add::dict_add_term(d_, multiply, pow(_base, self.get_exp()));
+            } else {
+                Add::dict_add_term(d_, multiply, self.rcp_from_this());
+            }
+            return;
+        }
+
+        mpz_class n = rcp_static_cast<const Integer>(self.get_exp())->as_mpz();
+        if (n < 0) {
+            return _coef_dict_add_term(multiply, div(one, expand(pow(_base, integer(-n)))));
+        }
+        RCP<const Add> base = rcp_static_cast<const Add>(_base);
+        umap_basic_num base_dict = base->dict_;
+        if (!(base->coef_->is_zero())) {
+            // Add the numerical coefficient into the dictionary. This
+            // allows a little bit easier treatment below.
+            insert(base_dict, base->coef_, one);
+        } else {
+            iaddnum(outArg(coeff), base->coef_);
+        }
+        if (n == 2) {
+            return square_expand(base_dict);
+        } else {
+            return pow_expand(base_dict, n.get_ui());
+        }
+    }
+
+    inline void _coef_dict_add_term(const RCP<const Number> &c, const RCP<const Basic> &term) {
+        if (is_a_Number(*term)) {
+            iaddnum(outArg(coeff), _mulnum(c, rcp_static_cast<const Number>(term)));
+        } else if (is_a<Add>(*term)) {
+            for (const auto &q: (rcp_static_cast<const Add>(term))->dict_)
+                Add::dict_add_term(d_, q.second, q.first);
+            iaddnum(outArg(coeff), rcp_static_cast<const Add>(term)->coef_);
+        } else {
+            RCP<const Number> coef2;
+            RCP<const Basic> t;
+            Add::as_coef_term(mul(c, term), outArg(coef2), outArg(t));
+            Add::dict_add_term(d_, coef2, t);
+        }
+    }
+};
+
+//! Expands `self`
+RCP<const Basic> expand(const RCP<const Basic> &self) {
+    ExpandVisitor v;
+    v.apply(self);
+    return Add::from_dict(v.coeff, std::move(v.d_));
```

Comment:
The last line should probably be a method in the `ExpandVisitor` and then `d_` and other variables should be private to the `ExpandVisitor` class.